phoenix-core/src/main/java/org/apache/phoenix/coprocessor/PhoenixTTLRegionObserver.java

```diff
@@ -0,0 +1,368 @@
+
+                    if (deleteIfExpired && deleteRowIfExpired(result)) {
+                        numRowsDeleted++;
+                        break;
+                    }
+                    // skip this row
+                    // 1. if the row has expired (checkRowNotExpired returned false)
+                    // 2. if the row was not deleted (deleteRowIfExpired returned false and
+                    //  do not want it to count towards the deleted count)
+                    if (maskIfExpired) {
+                        numRowsExpired++;
+                    }
+                    result.clear();
+                } while (hasMore);
+                return hasMore;
+            } catch (Throwable t) {
+                ServerUtil.throwIOException(region.getRegionInfo().getRegionNameAsString(), t);
+                return false; // impossible
+            }
+        }
+
+        /**
+         * @param cellList is an input and output parameter and will either include a valid row or be an empty list
+         * @return true if row expired and deleted or empty, otherwise false
+         * @throws IOException
+         */
+        private boolean deleteRowIfExpired(List<Cell> cellList) throws IOException {
+
+            long cellListSize = cellList.size();
+            if (cellListSize == 0) {
+                return true;
+            }
+
+            Iterator<Cell> cellIterator = cellList.iterator();
+            Cell firstCell = cellIterator.next();
+            byte[] rowKey = new byte[firstCell.getRowLength()];
+            System.arraycopy(firstCell.getRowArray(), firstCell.getRowOffset(), rowKey, 0,
+                    firstCell.getRowLength());
+
+            boolean isRowExpired = !checkRowNotExpired(cellList);
+            if (isRowExpired) {
+                long ttl = ScanUtil.getPhoenixTTL(this.scan);
+                long ts = getMaxTimestamp(cellList);
+                LOG.debug(String.format(
+                        "** PHOENIX-TTL: Deleting region = %s, row = %s, delete-ts = %d, max-ts = %d ** ",
+                        region.getRegionInfo().getTable().getNameAsString(), Bytes.toString(rowKey),
+                        now - ttl, ts));
+                Delete del = new Delete(rowKey, now - ttl);
+                Mutation[] mutations = new Mutation[] { del };
+                region.batchMutate(mutations, HConstants.NO_NONCE, HConstants.NO_NONCE);
+                return true;
+            }
+            return false;
+        }
+
+        private boolean isEmptyColumn(Cell cell) {
+            return Bytes.compareTo(cell.getFamilyArray(), cell.getFamilyOffset(),
+                    cell.getFamilyLength(), emptyCF, 0, emptyCF.length) == 0 &&
+                    Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(),
+                            cell.getQualifierLength(), emptyCQ, 0, emptyCQ.length) == 0;
+        }
+
+        // TODO : Remove it after we verify all SQLs include the empty column.
```

Comment:
Is this a left-over TODO, or do we need a follow-up Jira?

Reply:
Will create a follow-up Jira